### PR TITLE
Iothub metrics

### PR DESCRIPTION
**Pull request by maple10001**
# Description and Motivation 
Allow user fetch iothub metrics from device simulation service.

# Change type 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [x] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/250)

### CODE REVIEW

File: WebService/appsettings.ini [hunk 1]
```diff
@@ -31,6 +31,12 @@ webservice_url = "${PCS_STORAGEADAPTER_WEBSERVICE_URL}"
 # Max time allowed for requests to the storage adapter before reporting an error
 webservice_timeout = 10000
 
+[AzureManagementService]
+# URL where the azure management api is available, e.g. http://localhost:9022/v1 on local dev environments
```
Comment (hathind-ms) [inline WebService/appsettings.ini]: is this correct? Update comment

File: WebService/appsettings.ini [hunk 2]
```diff
@@ -31,6 +31,12 @@ webservice_url = "${PCS_STORAGEADAPTER_WEBSERVICE_URL}"
...
+webservice_url = "https://management.azure.com"
+
+# Max time allowed for requests to the azure management adapter before reporting an error
```
Comment (hathind-ms) [inline WebService/appsettings.ini]: replace adapter with api. "ax time allowed for requests to the azure management api before reporting an error"

File: WebService/v1/Controllers/SimulationsController.cs [hunk 3]
```diff
@@ -92,6 +97,17 @@ public async Task<SimulationApiModel> GetAsync(string id)
...
 
+        [HttpPost("{id}/metrics/iothub")]
+        public async Task<object> PostAsync(
+            [FromBody] MetricsRequestsApiModel requests,
+            [FromHeader]string authorization,
+            string id)
+        {
+            return requests == null
+                ? await this.iothubMetrics.GetIothubMetrics(authorization, null)
+                : await this.iothubMetrics.GetIothubMetrics(authorization, requests.ToServiceModel());
```
Comment (hathind-ms) [inline WebService/v1/Controllers/SimulationsController.cs]: can you try this:

var requestsModel = requests?.ToServiceModel();
await this.iothubMetrics.GetIothubMetrics(authorization, requestsModel);

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+{
+    public interface IAzureManagementAdapterClient
+    {
+        Task<MetricsResponseModel> PostAsync(string token, MetricsRequestsModel requests);
+    }
+
+    public class AzureManagementAdapterClient : IAzureManagementAdapterClient
+    {
+        private const string DATE_FORMAT = "yyyy-MM-ddTHH:mm:ssZ";
+        private const string METIRCS_API_VERSION = "2016-06-01";
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: typo: METRICS_API_VERSION
Should this be read from appsettings?

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        public async Task<MetricsResponseModel> PostAsync(string token, MetricsRequestsModel requests)
+        {
+            if (requests == null)
+            {
+                requests = this.GetMetricsRequests();
+            }
+
+            var request = this.PrepareRequest("batch?api-version=2015-11-01", token, requests);
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: why are you hardcoding api version here?

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            var response = await this.httpClient.PostAsync(request);
+
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: If there is error then we should log to Diagnostics as well.

File: WebService/v1/Controllers/SimulationsController.cs [hunk 7]
```diff
@@ -73,9 +76,11 @@ public async Task<SimulationApiModel> GetAsync(string id)
             [FromBody] SimulationApiModel simulationApiModel,
             [FromQuery(Name = "template")] string template = "")
         {
-            await simulationApiModel?.ValidateInputRequestAsync(this.log, this.connectionStringManager);
-
-            if (simulationApiModel == null)
+            if (simulationApiModel != null)
+            {
```
Comment (maple10001) [inline WebService/v1/Controllers/SimulationsController.cs]: Remove `?`

File: WebService/v1/Controllers/SimulationsController.cs [hunk 8]
```diff
@@ -92,12 +97,23 @@ public async Task<SimulationApiModel> GetAsync(string id)
                 simulation, this.servicesConfig, this.deploymentConfig, this.connectionStringManager, this.simulationRunner, this.rateReporter);
         }
 
+        [HttpPost("{id}/metrics/iothub")]
+        public async Task<object> PostAsync(
+            [FromBody] MetricsRequestsApiModel requests,
+            [FromHeader]string authorization,
```
Comment (maple10001) [inline WebService/v1/Controllers/SimulationsController.cs]: space

Comment (maple10001) [inline WebService/v1/Controllers/SimulationsController.cs]: never log token

File: WebService/v1/Controllers/SimulationsController.cs [hunk 9]
```diff
@@ -92,12 +97,23 @@ public async Task<SimulationApiModel> GetAsync(string id)
...
+            string id)
+        {
+            var payload = requests?.ToServiceModel();
```
Comment (maple10001) [inline WebService/v1/Controllers/SimulationsController.cs]: Add a comment about auto generated query

File: WebService/v1/Controllers/SimulationsController.cs [hunk 10]
```diff
@@ -92,12 +97,23 @@ public async Task<SimulationApiModel> GetAsync(string id)
...
+
+            return await this.iothubMetrics.GetIothubMetrics(authorization, payload);
+        }
+
         [HttpPut("{id}")]
         public async Task<SimulationApiModel> PutAsync(
             [FromBody] SimulationApiModel simulationApiModel,
             string id = "")
         {
-            await simulationApiModel?.ValidateInputRequestAsync(this.log, this.connectionStringManager);
```
Comment (maple10001) [inline WebService/v1/Controllers/SimulationsController.cs]: Reverit

File: Services/IotHub/IothubMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        public IotHubMetrics(IAzureManagementAdapterClient azureManagementAdapter)
+        {
+            this.azureManagementAdapter = azureManagementAdapter;
+        }
+
+        /// <summary>
+        /// Query Azure management API for the iothub metrics.
+        /// </summary>
+        /// <returns>Responses from the Azure management API</returns>
+        public async Task<MetricsResponseModel> GetIothubMetrics(string token, MetricsRequestsModel requests)
```
Comment (maple10001) [inline Services/IotHub/IothubMetrics.cs]: Add `Async`

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        public async Task<MetricsResponseModel> PostAsync(string token, MetricsRequestsModel requests)
+        {
+            if (requests == null)
+            {
+                requests = this.GetMetricsRequests();
+            }
+
+            var request = this.PrepareRequest($"batch?api-version={this.config.AzureManagementAdapterApiVersion}", token, requests);
+            var response = await this.httpClient.PostAsync(request);
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: try..catch

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                $"({string.Join(" or ", typeArray)})",
+                $"startTime eq {startTime}",
+                $"endTime eq {endTime}",
+                "timeGrain eq duration'PT1M'"
+            };
+            
+            return string.Join(" and ", filterArray);
+        }
+
+        private MetricsRequestsModel GetMetricsRequests()
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: rename to default

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.AzureManagementAdapter
+{
+    public interface IAzureManagementAdapterClient
+    {
+        Task<MetricsResponseModel> PostAsync(string token, MetricsRequestsModel requests);
+    }
+
+    public class AzureManagementAdapterClient : IAzureManagementAdapterClient
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: align with file name

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        private HttpRequest PrepareRequest(string path, string token, MetricsRequestsModel content = null)
+        {
+            var request = new HttpRequest();
+            request.AddHeader(HttpRequestHeader.Accept.ToString(), "application/json");
+            request.AddHeader(HttpRequestHeader.CacheControl.ToString(), "no-cache");
+            request.AddHeader(HttpRequestHeader.Authorization.ToString(), token);
+            request.SetUriFromString($"{this.config.AzureManagementAdapterApiUrl}/{path}");
+            request.Options.EnsureSuccess = false;
+            request.Options.Timeout = this.config.AzureManagementAdapterApiTimeout;
+            if (this.config.AzureManagementAdapterApiUrl.ToLowerInvariant().StartsWith("https:"))
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: verify it's contains https, throw exception

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            if (response.IsError)
+            {
+                this.diagnosticsLogger.LogServiceErrorAsync("Metrics request error", new { response.Content });
+                throw new ExternalDependencyException(
+                    new HttpRequestException($"Metrics request error: status code {response.StatusCode}"));
+            }
+        }
+
+        private string GetIoTHubMetricsUrl()
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: rename

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            return $"/subscriptions/{this.deploymentConfig.AzureSubscriptionId}" +
+                   $"/resourceGroups/{this.deploymentConfig.AzureResourceGroup}" +
+                   $"/providers/Microsoft.Devices/IotHubs/{this.deploymentConfig.AzureIothubName}" +
+                   $"/providers/Microsoft.Insights/metrics?api-version={METRICS_API_VERSION}&" +
+                   $"$filter={this.GetMetricsQuery()}";
+        }
+
+        private string GetMetricsQuery()
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: rename

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            // TODO: consider using query params from controller
+            var now = DateTimeOffset.UtcNow;
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: add infos to docs

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            if (requests == null)
+            {
+                requests = this.GetMetricsRequests();
+            }
+
+            var request = this.PrepareRequest($"batch?api-version={this.config.AzureManagementAdapterApiVersion}", token, requests);
+            var response = await this.httpClient.PostAsync(request);
+
+            this.log.Debug("Azure Management response", () => new { response });
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: add log for `requestURL`

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+            this.log.Debug("Azure Management request", () => new { request });
+
+            return request;
+        }
+
+        private void ThrowIfError(IHttpResponse response)
+        {
+            if (response.IsError)
+            {
+                this.diagnosticsLogger.LogServiceErrorAsync("Metrics request error", new { response.Content });
```
Comment (maple10001) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: add log

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs (hunk 12 above)
Comment (maple10001) [reply Services/AzureManagementAdapter/AzureManagementAdapter.cs]: `SendAysnc` in `HttpClient.cs` will transform `exception` to `new HttpResponse()`.

File: WebService/appsettings.ini [hunk 21]
```diff
@@ -31,6 +31,16 @@ webservice_url = "${PCS_STORAGEADAPTER_WEBSERVICE_URL}"
 # Max time allowed for requests to the storage adapter before reporting an error
 webservice_timeout = 10000
 
+[AzureManagementService]
+# URL where the azure management api is available.
+webservice_url = "https://management.azure.com"
+
+# Max time allowed for requests to the azure management service before reporting an error.
+# Value in milliseconds
+webservice_timeout = 10000
```
Comment (avanikp) [inline WebService/appsettings.ini]: this is already defined in line 32

Comment (hathind-ms) [reply WebService/appsettings.ini]: this value is specific to AzureManagementService calls

File: Services/AzureManagementAdapter/MetricsModel.cs [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    public class MetricDataModel
+    {
+        [JsonProperty("timeStamp")]
+        public string TimeStamp { get; set; }
+
+        [JsonProperty("total")]
+        public long Total { get; set; }
+    }
+
+    public class MetricsRequestsModel
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/MetricsModel.cs]: rename to MetricsRequestListModel

File: Services/AzureManagementAdapter/MetricsModel.cs [hunk 23]
```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.AzureManagementAdapter
+{
+    public class MetricsResponseModel
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/MetricsModel.cs]: rename to MetricsResponseListModel

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+        ///     https://docs.microsoft.com/en-us/rest/api/monitor/metrics/list
+        /// </summary>
+        /// <param name="token"></param>
+        /// <param name="requests"></param>
+        /// <returns></returns>
+        public async Task<MetricsResponseModel> PostAsync(MetricsRequestsModel requests)
+        {
+            if (this.AccessTokenIsNullOrEmpty()) await this.GetAadTokenAsync();
+
+            if (this.AccessTokenExpireSoon()) this.GetAadTokenAsync();
```
Comment (hathind-ms) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: Add a comment to explain the scenario

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+
+            if (requests == null) requests = this.GetDefaultMetricsRequests();
```
Comment (avanikp) [inline Services/AzureManagementAdapter/AzureManagementAdapter.cs]: I recommend keeping the API for getting metrics separate from AzureManagementAdapter. In future we will add additional APIs/requests which will use azure mgmt and will not want to set metrics request as default.

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+            this.log.Debug("Azure Management response", () => new { response });
+
+            this.ThrowIfError(response);
+
+            return JsonConvert.DeserializeObject<MetricsResponseModel>(response.Content);
+        }
+
+        private bool AccessTokenIsNullOrEmpty()
+        {
+            return this.secureAccessToken.Length == 0;
```
[the remark this answers is not in the document]
Comment (maple10001) [reply Services/AzureManagementAdapter/AzureManagementAdapter.cs]: It's been init as `new SecureString()` in constructor.

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs (hunk 25 above)
Comment (maple10001) [reply Services/AzureManagementAdapter/AzureManagementAdapter.cs]: This `PostAsync` method is for fetching metrics only. Please consider build other methods for different senarios.

File: WebService.Test/v1/Models/SimulationApiModel/MetricsApiModelTest.cs [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace WebService.Test.v1.Models.SimulationApiModel
+{
+    public class MetricsApiModelTest
+    {
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItReturnsMetricsRequestsModelFromApiModel()
+        {
+            // Arrange
+            var apiModel = new MetricsRequestsApiModel();
```
Comment (peterfelts) [inline WebService.Test/v1/Models/SimulationApiModel/MetricsApiModelTest.cs]: This test should add some requests to the API model, then confirm that they were converted properly.

File: WebService/appsettings.ini [hunk 28]
```diff
@@ -31,6 +31,32 @@ webservice_url = "${PCS_STORAGEADAPTER_WEBSERVICE_URL}"
...
+
+# Api version
+api_version = "2015-11-01"
```
Comment (peterfelts) [inline WebService/appsettings.ini]: I'd like to see these prefixed with something indicating that these configs are related to the AzureManagementService, something like: azure_management_webservice_url, etc. When I 'webservice_url', it isn't immediately clear to me that you're referring to the URL of the Azure Management service.

File: WebService/v1/Controllers/SimulationsController.cs [hunk 29]
```diff
@@ -95,19 +100,31 @@ public async Task<SimulationApiModel> GetAsync(string id)
                 simulation, this.servicesConfig, this.deploymentConfig, this.connectionStringManager, this.simulationRunner, this.rateReporter);
         }
 
+        [HttpPost("{id}/metrics/iothub!search")]
+        public async Task<object> PostAsync(
+            [FromBody] MetricsRequestsApiModel requests,
+            string id)
+        {
+            var payload = requests?.ToServiceModel();
```
Comment (peterfelts) [inline WebService/v1/Controllers/SimulationsController.cs]: We don't need to validate the received API model before processing it?

File: WebService/v1/Models/SimulationApiModel/MetricsApiModel.cs [hunk 30]
```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.AzureManagementAdapter;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.WebService.v1.Models.SimulationApiModel
+{
+    public class MetricsRequestApiModel
```
Comment (peterfelts) [inline WebService/v1/Models/SimulationApiModel/MetricsApiModel.cs]: The class name and file name should match.

File: WebService/v1/Models/SimulationApiModel/MetricsApiModel.cs [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            return new MetricsRequestModel()
+            {
+                HttpMethod = this.HttpMethod,
+                RelativeUrl = this.RelativeUrl
+            };
+        }
+    }
+
+    public class MetricsRequestsApiModel
```
Comment (peterfelts) [inline WebService/v1/Models/SimulationApiModel/MetricsApiModel.cs]: Please put these in two different files.

File: Services/AzureManagementAdapter/AzureManagementAdapter.cs [hunk 32]
```diff
@@ -0,0 +1,239 @@
...
+        private string GetDefaultIoTHubMetricsUrl()
+        {
+            return $"/subscriptions/{this.deploymentConfig.AzureSubscriptionId}" +
+                   $"/resourceGroups/{this.deploymentConfig.AzureResourceGroup}" +
+                   $"/providers/Microsoft.Devices/IotHubs/{this.deploymentConfig.AzureIothubName}" +
+                   $"/providers/Microsoft.Insights/metrics?api-version={METRICS_API_VERSION}&" +
+                   $"$filter={this.GetDefaultMetricsQuery()}";
+        }
+
+        private async Task GetAadTokenAsync()
```
[the remark this answers is not in the document]
Comment (maple10001) [reply Services/AzureManagementAdapter/AzureManagementAdapter.cs]: Will consider this approch when we adopt `Azure Key Vault`. Added todo inline.

File: WebService.Test/v1/Controllers/SimulationsControllerTest.cs [hunk 33]
```diff
@@ -0,0 +1,294 @@
...
+        private readonly SimulationsController target;
+
+        private Simulation GetSimulationById(string id)
+        {
+            return new Simulation
+            {
+                Id = id,
+                DeviceModels = new List<Simulation.DeviceModelRef>
+                {
+                    new Simulation.DeviceModelRef { Id = "Chiller_01", Count = 10 }
```
[the remark this answers is not in the document]
Comment (maple10001) [reply WebService.Test/v1/Controllers/SimulationsControllerTest.cs]: Will consider adding more type in the integration tests.

File: WebService.Test/v1/Controllers/SimulationsControllerTest.cs [hunk 34]
```diff
@@ -0,0 +1,294 @@
...
+                    this.deploymentConfig.Object,
+                    this.connectionStringManager.Object,
+                    this.simulationRunner.Object,
+                    this.rateReporter.Object
+                )
+            ).Result;
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.Equal(simulation.Id, result.Id);
```
[the remark this answers is not in the document]
Comment (maple10001) [reply WebService.Test/v1/Controllers/SimulationsControllerTest.cs]: There are tests for model transformation inside `simulationApimodelTests`.

File: WebService.Test/v1/Controllers/SimulationsControllerTest.cs [hunk 35]
```diff
@@ -0,0 +1,294 @@
...
+
+            this.simulationsService
+                .Setup(x => x.GetListAsync())
+                .ReturnsAsync(simulations);
+
+            // Act
+            var result = this.target.GetAsync().Result;
+
+            // Assert
+            Assert.Equal(simulations.Count, result.Items.Count);
```
[the remark this answers is not in the document]
Comment (maple10001) [reply WebService.Test/v1/Controllers/SimulationsControllerTest.cs]: Part of the model tests.

File: WebService/appsettings.ini (hunk 28 above)
Comment (maple10001) [reply WebService/appsettings.ini]: Please refer to `/WebService/Runtime/Config.cs`.

File: WebService/v1/Controllers/SimulationsController.cs (hunk 29 above)
Comment (maple10001) [reply WebService/v1/Controllers/SimulationsController.cs]: No need for crawl.

Comment (peterfelts) [reply WebService/v1/Controllers/SimulationsController.cs]: Nit: "API payload validation is not required as we're simply relaying the request."

**Pull request merged**